File: services/knowledge-builder/src/builder/chunker.ts
import crypto from 'crypto'
import type { HeadingItem } from '@actionbookdev/db'

export interface ChunkerOptions {
  chunkSize: number // Target tokens per chunk
  chunkOverlap: number // Overlap tokens between chunks
  minChunkSize: number // Minimum chunk size
  splitHeadingLevel: number // Only split at this heading level (1=H1, 2=H2, etc.)
}

export interface ChunkData {
  content: string
  chunkIndex: number
  startChar: number
  endChar: number
  heading?: string
  headingHierarchy: HeadingItem[]
  tokenCount: number
}

const DEFAULT_OPTIONS: ChunkerOptions = {
  chunkSize: 2000,
  chunkOverlap: 50,
  minChunkSize: 100,
  splitHeadingLevel: 2,
}

interface Section {
  content: string
  heading: string
  headingHierarchy: HeadingItem[]
  startChar: number
}

export class DocumentChunker {
  private options: ChunkerOptions

  constructor(options: Partial<ChunkerOptions> = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options }
  }

  // Headings to skip (noise sections that don't contain useful content)
  private static SKIP_HEADINGS = [
    'Related articles',
    'Related topics',
    'Site Footer',
    'Support',
    'Hosting',
    'Airbnb',
  ]

  /**
   * Split a markdown document into chunks
   */
  chunk(content: string): ChunkData[] {
    // If chunkSize is very large (e.g., --no-chunk mode), return entire content as single chunk
    const contentTokens = this.estimateTokens(content)
    if (contentTokens <= this.options.chunkSize) {
      const h1Match = content.match(/^#\s+(.+)$/m)
      return [
        {
          content: content.trim(),
          chunkIndex: 0,
          startChar: 0,
          endChar: content.length,
          heading: h1Match ? h1Match[1].trim() : '',
          headingHierarchy: h1Match
            ? [{ level: 1, text: h1Match[1].trim() }]
            : [],
          tokenCount: contentTokens,
        },
      ]
    }

    const sections = this.splitByHeadings(content)
    const chunks: ChunkData[] = []
    let chunkIndex = 0

    for (const section of sections) {
      // Skip noise sections like "Related articles"
      if (DocumentChunker.SKIP_HEADINGS.includes(section.heading)) {
        continue
      }

      const sectionChunks = this.chunkSection(section, chunkIndex)
      chunks.push(...sectionChunks)
      chunkIndex += sectionChunks.length
    }

    return chunks
  }

  /**
   * Split content by markdown headings at the specified level
   */
  private splitByHeadings(content: string): Section[] {
    const splitLevel = this.options.splitHeadingLevel
    const headingPattern = new RegExp(`^(#{1,${splitLevel}})\\s+(.+)$`, 'gm')
    const headingStack: Array<{ level: number; text: string }> = []

    // Find all split positions
    headingPattern.lastIndex = 0
    const splitPositions: number[] = []
    let match

    while ((match = headingPattern.exec(content)) !== null) {
      if (match[1].length === splitLevel) {
        splitPositions.push(match.index)
      }
    }
    splitPositions.push(content.length)

    // Rebuild sections with content
    const finalSections: Section[] = []
    headingPattern.lastIndex = 0
    headingStack.length = 0

    for (let i = 0; i < splitPositions.length - 1; i++) {
      const start = splitPositions[i]
      const end = splitPositions[i + 1]
      const sectionContent = content.substring(start, end)

      // Extract heading from this section
      const firstLineMatch = sectionContent.match(/^(#{1,6})\s+(.+)$/m)
      const heading = firstLineMatch ? firstLineMatch[2].trim() : ''
      const level = firstLineMatch ? firstLineMatch[1].length : splitLevel

      // Update heading stack
      while (
        headingStack.length > 0 &&
        headingStack[headingStack.length - 1].level >= level
      ) {
        headingStack.pop()
      }
      if (heading) {
        headingStack.push({ level, text: heading })
      }

      finalSections.push({
        content: sectionContent,
        heading,
        headingHierarchy: headingStack.map((h) => ({
          level: h.level,
          text: h.text,
        })),
        startChar: start,
      })
    }

    // Handle content before first split-level heading
    if (splitPositions[0] > 0) {
      const introContent = content.substring(0, splitPositions[0])
      if (introContent.trim()) {
        const h1Match = introContent.match(/^#\s+(.+)$/m)
        finalSections.unshift({
          content: introContent,
          heading: h1Match ? h1Match[1].trim() : '',
          headingHierarchy: h1Match
            ? [{ level: 1, text: h1Match[1].trim() }]
            : [],
          startChar: 0,
        })
      }
    }

    // If no sections found, treat entire content as one section
    if (finalSections.length === 0) {
      const h1Match = content.match(/^#\s+(.+)$/m)
      finalSections.push({
        content,
        heading: h1Match ? h1Match[1].trim() : '',
        headingHierarchy: h1Match
          ? [{ level: 1, text: h1Match[1].trim() }]
          : [],
        startChar: 0,
      })
    }

    return finalSections
  }

  /**
   * Chunk a single section
   */
  private chunkSection(section: Section, startIndex: number): ChunkData[] {
    const tokens = this.estimateTokens(section.content)

    // If section is small enough, return as single chunk
    if (tokens <= this.options.chunkSize) {
      return [
        {
          content: section.content.trim(),
          chunkIndex: startIndex,
          startChar: section.startChar,
          endChar: section.startChar + section.content.length,
          heading: section.heading,
          headingHierarchy: section.headingHierarchy,
          tokenCount: tokens,
        },
      ]
    }

    // Split by paragraphs
    const paragraphs = this.splitParagraphs(section.content)
    const chunks: ChunkData[] = []

    let currentChunk = ''
    let currentTokens = 0
    let chunkStartChar = section.startChar
    let charOffset = 0

    for (const paragraph of paragraphs) {
      const paraTokens = this.estimateTokens(paragraph)

      // If adding this paragraph exceeds chunk size
      if (currentTokens + paraTokens > this.options.chunkSize && currentChunk) {
        // Save current chunk
        chunks.push({
          content: currentChunk.trim(),
          chunkIndex: startIndex + chunks.length,
          startChar: chunkStartChar,
          endChar: section.startChar + charOffset,
          heading: section.heading,
          headingHierarchy: section.headingHierarchy,
          tokenCount: currentTokens,
        })

        // Start new chunk with overlap
        const overlap = this.getOverlapText(currentChunk)
        currentChunk = overlap + paragraph
        currentTokens = this.estimateTokens(currentChunk)
        chunkStartChar = section.startChar + charOffset - overlap.length
      } else {
        currentChunk += paragraph
        currentTokens += paraTokens
      }

      charOffset += paragraph.length
    }

    // Add final chunk - merge into previous if too small to preserve content
    if (currentChunk.trim()) {
      const finalTokens = this.estimateTokens(currentChunk)

      if (finalTokens >= this.options.minChunkSize || chunks.length === 0) {
        // Normal case: add as new chunk
        chunks.push({
          content: currentChunk.trim(),
          chunkIndex: startIndex + chunks.length,
          startChar: chunkStartChar,
          endChar: section.startChar + section.content.length,
          heading: section.heading,
          headingHierarchy: section.headingHierarchy,
          tokenCount: finalTokens,
        })
      } else {
        // Small final chunk: merge into previous chunk to avoid losing content
        const lastChunk = chunks[chunks.length - 1]
        lastChunk.content = lastChunk.content + '\n\n' + currentChunk.trim()
        lastChunk.endChar = section.startChar + section.content.length
        lastChunk.tokenCount = this.estimateTokens(lastChunk.content)
      }
    }

    return chunks
  }

  /**
   * Split content into paragraphs while preserving code blocks
   */
  private splitParagraphs(content: string): string[] {
    // Protect code blocks
    const codeBlocks: string[] = []
    const protectedContent = content.replace(/```[\s\S]*?```/g, (match) => {
      codeBlocks.push(match)
      return `__CODE_BLOCK_${codeBlocks.length - 1}__`
    })

    // Split by double newlines
    const paragraphs = protectedContent.split(/\n\s*\n/).map((p) => p + '\n\n')

    // Restore code blocks
    return paragraphs.map((para) => {
      return para.replace(/__CODE_BLOCK_(\d+)__/g, (_, index) => {
        return codeBlocks[parseInt(index)]
      })
    })
  }

  /**
   * Get overlap text from the end of a chunk
   */
  private getOverlapText(text: string): string {
    const words = text.trim().split(/\s+/)
    const overlapWords = words.slice(-this.options.chunkOverlap)
    return overlapWords.join(' ') + ' '
  }

  /**
   * Estimate token count (rough approximation)
   */
  private estimateTokens(text: string): number {
    return Math.ceil(text.length / 4)
  }
}

/**
 * Generate a hash for chunk content
 */
export function hashChunk(content: string): string {
  return crypto
    .createHash('sha256')
    .update(content)
    .digest('hex')
    .substring(0, 16)
}

File: services/knowledge-builder-any/src/chunker.rs
//! Document chunker for splitting markdown content into smaller pieces
//!
//! Ported from TypeScript implementation in services/knowledge-builder

use crate::db::models::HeadingItem;
use regex::Regex;
use sha2::{Digest, Sha256};

/// Chunker configuration options
#[derive(Debug, Clone)]
pub struct ChunkerOptions {
    /// Target tokens per chunk
    pub chunk_size: usize,
    /// Overlap tokens between chunks
    pub chunk_overlap: usize,
    /// Minimum chunk size
    pub min_chunk_size: usize,
    /// Only split at this heading level (1=H1, 2=H2, etc.)
    pub split_heading_level: usize,
}

impl Default for ChunkerOptions {
    fn default() -> Self {
        Self {
            chunk_size: 2000,
            chunk_overlap: 50,
            min_chunk_size: 100,
            split_heading_level: 2,
        }
    }
}

/// Data for a single chunk
#[derive(Debug, Clone)]
pub struct ChunkData {
    pub content: String,
    pub chunk_index: i32,
    pub start_char: i32,
    pub end_char: i32,
    pub heading: Option<String>,
    pub heading_hierarchy: Vec<HeadingItem>,
    pub token_count: i32,
}

/// Internal section representation
#[derive(Debug)]
struct Section {
    content: String,
    heading: String,
    heading_hierarchy: Vec<HeadingItem>,
    start_char: usize,
}

/// Headings to skip (noise sections)
const SKIP_HEADINGS: &[&str] = &[
    "Related articles",
    "Related topics",
    "Site Footer",
    "Support",
    "Hosting",
    "Airbnb",
];

/// Document chunker
pub struct DocumentChunker {
    options: ChunkerOptions,
}

impl DocumentChunker {
    /// Create a new chunker with options
    pub fn new(options: ChunkerOptions) -> Self {
        Self { options }
    }

    /// Split a markdown document into chunks
    pub fn chunk(&self, content: &str) -> Vec<ChunkData> {
        let sections = self.split_by_headings(content);
        let mut chunks = Vec::new();
        let mut chunk_index = 0;

        for section in sections {
            // Skip noise sections
            if SKIP_HEADINGS.contains(&section.heading.as_str()) {
                continue;
            }

            let section_chunks = self.chunk_section(&section, chunk_index);
            chunk_index += section_chunks.len() as i32;
            chunks.extend(section_chunks);
        }

        chunks
    }

    /// Split content by markdown headings at the specified level
    fn split_by_headings(&self, content: &str) -> Vec<Section> {
        let split_level = self.options.split_heading_level;

        // Pattern to match headings up to split_level
        let pattern = format!(r"(?m)^(#{{1,{}}})[\t ]+(.+)$", split_level);
        let heading_re = Regex::new(&pattern).unwrap();

        // Find all split positions (where heading level equals split_level)
        let mut split_positions: Vec<usize> = Vec::new();

        for cap in heading_re.captures_iter(content) {
            let hashes = cap.get(1).unwrap().as_str();
            if hashes.len() == split_level {
                split_positions.push(cap.get(0).unwrap().start());
            }
        }
        split_positions.push(content.len());

        let mut sections = Vec::new();
        let mut heading_stack: Vec<(usize, String)> = Vec::new();

        // Handle content before first split-level heading
        if !split_positions.is_empty() && split_positions[0] > 0 {
            let intro_content = &content[0..split_positions[0]];
            if !intro_content.trim().is_empty() {
                let h1_re = Regex::new(r"(?m)^#[\t ]+(.+)$").unwrap();
                let (heading, hierarchy) = if let Some(cap) = h1_re.captures(intro_content) {
                    let text = cap.get(1).unwrap().as_str().trim().to_string();
                    (text.clone(), vec![HeadingItem { level: 1, text }])
                } else {
                    (String::new(), vec![])
                };

                sections.push(Section {
                    content: intro_content.to_string(),
                    heading,
                    heading_hierarchy: hierarchy,
                    start_char: 0,
                });
            }
        }

        // Process each section
        for i in 0..split_positions.len().saturating_sub(1) {
            let start = split_positions[i];
            let end = split_positions[i + 1];
            let section_content = &content[start..end];

            // Extract heading from this section
            let first_line_re = Regex::new(r"(?m)^(#{1,6})[\t ]+(.+)$").unwrap();
            let (heading, level) = if let Some(cap) = first_line_re.captures(section_content) {
                let hashes = cap.get(1).unwrap().as_str();
                let text = cap.get(2).unwrap().as_str().trim().to_string();
                (text, hashes.len())
            } else {
                (String::new(), split_level)
            };

            // Update heading stack
            while !heading_stack.is_empty() && heading_stack.last().unwrap().0 >= level {
                heading_stack.pop();
            }
            if !heading.is_empty() {
                heading_stack.push((level, heading.clone()));
            }

            sections.push(Section {
                content: section_content.to_string(),
                heading,
                heading_hierarchy: heading_stack
                    .iter()
                    .map(|(l, t)| HeadingItem {
                        level: *l as i32,
                        text: t.clone(),
                    })
                    .collect(),
                start_char: start,
            });
        }

        // If no sections found, treat entire content as one section
        if sections.is_empty() {
            let h1_re = Regex::new(r"(?m)^#[\t ]+(.+)$").unwrap();
            let (heading, hierarchy) = if let Some(cap) = h1_re.captures(content) {
                let text = cap.get(1).unwrap().as_str().trim().to_string();
                (text.clone(), vec![HeadingItem { level: 1, text }])
            } else {
                (String::new(), vec![])
            };

            sections.push(Section {
                content: content.to_string(),
                heading,
                heading_hierarchy: hierarchy,
                start_char: 0,
            });
        }

        sections
    }

    /// Chunk a single section
    fn chunk_section(&self, section: &Section, start_index: i32) -> Vec<ChunkData> {
        let tokens = self.estimate_tokens(&section.content);

        // If section is small enough, return as single chunk
        if tokens <= self.options.chunk_size {
            return vec![ChunkData {
                content: section.content.trim().to_string(),
                chunk_index: start_index,
                start_char: section.start_char as i32,
                end_char: (section.start_char + section.content.len()) as i32,
                heading: if section.heading.is_empty() {
                    None
                } else {
                    Some(section.heading.clone())
                },
                heading_hierarchy: section.heading_hierarchy.clone(),
                token_count: tokens as i32,
            }];
        }

        // Split by paragraphs
        let paragraphs = self.split_paragraphs(&section.content);
        let mut chunks = Vec::new();

        let mut current_chunk = String::new();
        let mut current_tokens = 0;
        let mut chunk_start_char = section.start_char;
        let mut char_offset = 0;

        for paragraph in &paragraphs {
            let para_tokens = self.estimate_tokens(paragraph);

            // If adding this paragraph exceeds chunk size
            if current_tokens + para_tokens > self.options.chunk_size && !current_chunk.is_empty() {
                // Save current chunk
                chunks.push(ChunkData {
                    content: current_chunk.trim().to_string(),
                    chunk_index: start_index + chunks.len() as i32,
                    start_char: chunk_start_char as i32,
                    end_char: (section.start_char + char_offset) as i32,
                    heading: if section.heading.is_empty() {
                        None
                    } else {
                        Some(section.heading.clone())
                    },
                    heading_hierarchy: section.heading_hierarchy.clone(),
                    token_count: current_tokens as i32,
                });

                // Start new chunk with overlap
                let overlap = self.get_overlap_text(&current_chunk);
                current_chunk = format!("{}{}", overlap, paragraph);
                current_tokens = self.estimate_tokens(&current_chunk);
                chunk_start_char = section.start_char + char_offset - overlap.len();
            } else {
                current_chunk.push_str(paragraph);
                current_tokens += para_tokens;
            }

            char_offset += paragraph.len();
        }

        // Add final chunk
        if !current_chunk.trim().is_empty() {
            let final_tokens = self.estimate_tokens(&current_chunk);

            if final_tokens >= self.options.min_chunk_size || chunks.is_empty() {
                chunks.push(ChunkData {
                    content: current_chunk.trim().to_string(),
                    chunk_index: start_index + chunks.len() as i32,
                    start_char: chunk_start_char as i32,
                    end_char: (section.start_char + section.content.len()) as i32,
                    heading: if section.heading.is_empty() {
                        None
                    } else {
                        Some(section.heading.clone())
                    },
                    heading_hierarchy: section.heading_hierarchy.clone(),
                    token_count: final_tokens as i32,
                });
            } else if let Some(last_chunk) = chunks.last_mut() {
                // Merge small final chunk into previous
                last_chunk.content = format!("{}\n\n{}", last_chunk.content, current_chunk.trim());
                last_chunk.end_char = (section.start_char + section.content.len()) as i32;
                last_chunk.token_count = self.estimate_tokens(&last_chunk.content) as i32;
            }
        }

        chunks
    }

    /// Split content into paragraphs while preserving code blocks
    fn split_paragraphs(&self, content: &str) -> Vec<String> {
        // Protect code blocks
        let code_block_re = Regex::new(r"```[\s\S]*?```").unwrap();
        let mut code_blocks = Vec::new();
        let protected_content = code_block_re
            .replace_all(content, |caps: &regex::Captures| {
                code_blocks.push(caps[0].to_string());
                format!("__CODE_BLOCK_{}__", code_blocks.len() - 1)
            })
            .to_string();

        // Split by double newlines
        let paragraphs: Vec<String> = protected_content
            .split("\n\n")
            .map(|p| format!("{}\n\n", p))
            .collect();

        // Restore code blocks
        paragraphs
            .into_iter()
            .map(|para| {
                let code_block_placeholder_re = Regex::new(r"__CODE_BLOCK_(\d+)__").unwrap();
                code_block_placeholder_re
                    .replace_all(&para, |caps: &regex::Captures| {
                        let index: usize = caps[1].parse().unwrap();
                        code_blocks.get(index).cloned().unwrap_or_default()
                    })
                    .to_string()
            })
            .collect()
    }

    /// Get overlap text from the end of a chunk
    fn get_overlap_text(&self, text: &str) -> String {
        let words: Vec<&str> = text.trim().split_whitespace().collect();
        let start = words.len().saturating_sub(self.options.chunk_overlap);
        let overlap_words: Vec<&str> = words[start..].to_vec();
        format!("{} ", overlap_words.join(" "))
    }

    /// Estimate token count (rough approximation: ~4 chars per token)
    fn estimate_tokens(&self, text: &str) -> usize {
        (text.len() + 3) / 4 // Ceiling division
    }
}

impl Default for DocumentChunker {
    fn default() -> Self {
        Self::new(ChunkerOptions::default())
    }
}

/// Generate a hash for content (SHA256, first 16 chars)
pub fn hash_content(content: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(content.as_bytes());
    hex::encode(hasher.finalize())[..16].to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_chunk() {
        let chunker = DocumentChunker::default();
        let content = "# Hello\n\nThis is a test document.";
        let chunks = chunker.chunk(content);

        assert!(!chunks.is_empty());
        assert!(chunks[0].content.contains("Hello"));
    }

    #[test]
    fn test_hash_content() {
        let hash = hash_content("Hello, World!");
        assert_eq!(hash.len(), 16);

        // Same content should produce same hash
        let hash2 = hash_content("Hello, World!");
        assert_eq!(hash, hash2);
    }

    #[test]
    fn test_heading_extraction() {
        let chunker = DocumentChunker::default();
        let content = "# Main Title\n\n## Section 1\n\nContent 1\n\n## Section 2\n\nContent 2";
        let chunks = chunker.chunk(content);

        assert!(chunks.len() >= 2);
    }
}
